### PR TITLE
Ensure we test against numpy 2 in CI

### PR DESCRIPTION
**Pull request by jrbourbeau**
NumPy 2 came out over the weekend. We've been testing against the nightly version of `numpy` in our `upstream` build but due to other dependencies, `numpy >=2` isn't getting installed on our main CI environments. 

This PR removes some optional dependencies (e.g. `numba`) to ensure `numpy` 2 can be installed in our Python 3.12 builds.

### CODE REVIEW

**Comment by jrbourbeau** [comment]
Okay, so there are some doctest failures (I'll push up a fix for that) and a single test failure

```
FAILED dask/tests/test_tokenize.py::test_tokenize_random_functions[np.random] - AssertionError: assert '5a758251ee05...fadb5e734ae74' == 'f8c814ecbf02...12210d58c137e'
  
  - f8c814ecbf0209a0c4812210d58c137e
  + 5a758251ee0551802acfadb5e734ae74
```

@fjetter do you have a moment to look at that test failure? It was added in https://github.com/dask/dask/pull/10865, which you reviewed

File: dask/tests/test_tokenize.py [hunk 1]
```diff
@@ -1271,6 +1271,7 @@ def test_tokenize_random_functions(module):
     """random.random() and other methods of the global random state do not compare as
     equal to themselves after a pickle roundtrip"""
     module = eval(module)
+    module.seed(2)
```
Comment (jrbourbeau) [inline dask/tests/test_tokenize.py:1274]: Manually setting the random seed here gets this test passing. If we don't do this, then this test fails against NumPy 2 with this tokenization mismatch error:

```
__________________ test_tokenize_random_functions[np.random] ___________________
[gw0] linux -- Python 3.12.4 /home/runner/miniconda3/envs/test-environment/bin/python3.12

module = <module 'numpy.random' from '/home/runner/miniconda3/envs/test-environment/lib/python3.12/site-packages/numpy/random/__init__.py'>

    @pytest.mark.parametrize(
        "module",
        ["random", pytest.param("np.random", marks=pytest.mark.skipif("not np"))],
    )
    def test_tokenize_random_functions(module):
        """random.random() and other methods of the global random state do not compare as
        equal to themselves after a pickle roundtrip"""
        module = eval(module)
    
        a = module.random
        b = pickle.loads(pickle.dumps(a))
>       assert check_tokenize(a) == check_tokenize(b)
E       AssertionError: assert '5a758251ee05...fadb5e734ae74' == 'f8c814ecbf02...12210d58c137e'
E         
E         - f8c814ecbf0209a0c4812210d58c137e
E         + 5a758251ee0551802acfadb5e734ae74

dask/tests/test_tokenize.py:1277: AssertionError
```

@fjetter @phofl  @hendrikmakait do any of you have more context here?

File: continuous_integration/environment-3.12.yaml [hunk 2]
```diff
@@ -23,13 +23,15 @@ dependencies:
...
   - pandas
-  - numba
+  # Temporarily removing to allow `numpy >=2` to be installed
+  # - numba
   - flask
   - fastparquet
   - h5py
-  - pytables
+  # Temporarily removing to allow `numpy >=2` to be installed
+  # - pytables
```
Comment (jakirkham) [inline continuous_integration/environment-3.12.yaml:34]: PyTables will soon be installable with NumPy 2. Just working out some minor kinks: https://github.com/PyTables/PyTables/issues/1200

File: continuous_integration/environment-3.12.yaml [hunk 3]
```diff
@@ -61,11 +63,13 @@ dependencies:
   - lz4
   - psutil
   - requests
-  - scikit-image
+  # Temporarily removing to allow `numpy >=2` to be installed
+  # - scikit-image
```
Comment (jakirkham) [inline continuous_integration/environment-3.12.yaml:67]: As of scikit-image 0.23.0, NumPy 2 is supported: https://github.com/scikit-image/scikit-image/issues/7282

File: continuous_integration/environment-3.12.yaml [hunk 4]
```diff
@@ -61,11 +63,13 @@ dependencies:
...
   - scikit-learn
   - scipy
   - python-snappy
-  - sparse
+  # Temporarily removing to allow `numpy >=2` to be installed
+  # - sparse
```
Comment (jakirkham) [inline continuous_integration/environment-3.12.yaml:72]: This was done in sparse 0.15.2: https://github.com/pydata/sparse/issues/680

**Review by jakirkham — COMMENTED** [review]
Thanks James! 🙏

Included a few notes on dependencies available on NumPy 2, which includes most (and soon will be all) of the ones commented currently

File: dask/tests/test_tokenize.py (hunk 1 above)
Comment (phofl) [reply dask/tests/test_tokenize.py:1274]: not really no, but this sounds like an acceptable solution

**Review by phofl — APPROVED** [review]

**Comment by phofl** [comment]
thx

**Pull request merged**